### DISCUSSION TITLE
Xtstate log collection

### DISCUSSION THREAD

**Discussion by dsathyakumar** [statelyai/xstate#2085] [Q&A] [answered]
Hello 👋 

Does Xstate have any feature such as being able to log into an object and provide a mechanism to be able to collect all the logs and dispatch them?

Currently, the **LOG** action appears to log to console. 
But it would be beneficial if this were pushed to an internally maintained Log object which can then be queried for collection & dispatch

**Comment by davidkpiano** [comment, marked as answer]
Sure:

```js
const service = interpret(someMachine, {
  logger: (data) => {
    // do anything you want with the data; e.g., custom logger
    // replaces default console.log()
  }
});
```

**Reply by dsathyakumar** [reply]
Thanks @davidkpiano 👍 
This works. But is there any place we can store this on the machine itself ?
Perhaps access to context ?
This signature gets the data alone.

**Reply by davidkpiano** [reply]
There is not; logging is a side-effect.

If you really wanted to, you can make a custom `"LOG"` event and handle it as a top-level transition to `assign(...)` to context and append to logs.
